tests/ssr/utils.tsx

```diff
@@ -0,0 +1,37 @@
+import { render } from '@testing-library/react';
+import { renderToString } from 'react-dom/server';
+import { FunctionComponent } from 'react';
+
+const USE_LAYOUT_EFFECT_ERROR = /useLayoutEffect does nothing on the server/;
+const originalConsoleError = console.error.bind(console.error);
+
+export function testSSR(App: FunctionComponent) {
+  const renderOnServer = () => renderToString(<App />);
+  const container = document.body.appendChild(document.createElement('div'));
+  const consoleErrorSpy = jest.spyOn(global.console, 'error')
+    .mockImplementation((msg) => {
+      /**
+       * Caveat of SSR: useEffect and useLayoutEffect hooks do not run when
+       * rendering on server.
+       *
+       * Suppress useLayoutEffect warnings here since the mock window in jest test
+       * environment made the workaround with isomorphic-layout-effect ineffective.
+       */
+      if (!msg.toString().match(USE_LAYOUT_EFFECT_ERROR)) {
+        originalConsoleError(msg);
+      }
+    });
+
+  // server render components to static markup
+  container.innerHTML = renderOnServer();
+
+  // hydrate a container whose HTML contents were rendered by ReactDOMServer
+  render(<App />, { container, hydrate: true });
+  expect(consoleErrorSpy).not.toBeCalledWith(
+    expect.stringContaining('Warning: Prop `%s` did not match. Server: %s Client: %s%s'),
+    'id',
```

Comment:
don't need to use string concatenation for a hardcoded value

Reply:
removed the `expect.stringContaining`, the rest of the arguments is needed in order of the params received by console error.

Reply:
Here does the %s basically mean "any string"?

Reply:
I think here it's react itself that's calling console.error
If react changes the implementation of their console.error call this test will stop functioning properly right? is there some other way to test this?

Reply:
yea, all prop mismatched error will use this message format. I suppose we don't need to explicitly check for `id` mismatched only, although I don't see other ways of testing this beside checking console error. I attempted to check for the html content of server vs client but it didn't work -- the actual of the container after hydration uses the same as server even though the error is present.
I decided to update the test to assert the number of error logged outside of the useLayoutEffect warning.

Reply:
Sure that makes sense, I can't think of a better way to test this. If they switch to console.warn for example maybe we start getting false positives from the test, but we'd also notice console.warn's in the jest output (hopefully). If they remove the console messages then 🤷 but seems unlikely